services/gateway/security/src/main/java/org/eclipse/ditto/services/gateway/security/authentication/jwt/DittoPublicKeyProvider.java

```diff
@@ -178,22 +183,50 @@ private HttpResponse getPublicKeysFromDiscoveryEndpoint(final String discoveryEn
                     .thenCompose(jwksUri -> httpClient.createSingleHttpRequest(HttpRequest.GET(jwksUri)))
                     .toCompletableFuture()
                     .get(JWK_REQUEST_TIMEOUT_MILLISECONDS, TimeUnit.MILLISECONDS);
-        } catch (final ExecutionException | InterruptedException | TimeoutException e) {
-            Thread.currentThread().interrupt();
-            throw new IllegalStateException(
-                    MessageFormat.format("Got Exception from discovery endpoint <{0}>.", discoveryEndpoint), e);
+        } catch (final ExecutionException e) {
+            throw DittoRuntimeException.asDittoRuntimeException(e, cause -> {
```

Comment:
I'd say it depends on the point of view. It seems to me that the code as it was previously structured wants to throw an IllegalStateException. I just changed that behavior in case the cause was a DittoRuntimeException and that's what this now implemented behaviour does. It tries to extract a DittoRuntimeException from the cause, and if that's not possible because there is none, it passes the actual cause of the ExecutionException to the `alternativeExceptionBuilder` where I kept the previous behaviour and threw an IllegalStateException.

Reply:
Maybe we can add a comment here as it isn't obvious that this is the desired behavior.

Reply:
Hm I think if code needs such explanation it's bad code. What do you think about wrapping a non DittoRuntimeException into a something like a "PublicKeyProviderException" and then introduce a separate method which tries to extract the DittoRuntimeException from the ExecutionException if possible and throws this generic "PublicKeyProviderException" if not?
Seems to be more clear as it's a separate method with a name indicating what it should do.